akka-http-core/src/main/scala/akka/http/impl/engine/rendering/HttpResponseRendererFactory.scala

```diff
@@ -94,20 +94,26 @@ private[http] class HttpResponseRendererFactory(
           override def onUpstreamFinish(): Unit =
             if (transferring) closeMode = CloseConnection
             else completeStage()
+
+          override def onUpstreamFailure(ex: Throwable): Unit = {
+            stopTransfer()
```

Comment:
This is the fix. The rest is refactoring to make `stopTransfer` accessible here.